### ISSUE TITLE
pyasn1==0.3.1, pyasn-modules==0.0.9 incompatible

### ISSUE THREAD

**Issue by cmcarthur** [snowflakedb/snowflake-connector-python#27]
pyasn1 v0.3.1 was released today, and it changes the constructor APIs of some underlying types:

https://github.com/etingof/pyasn1/blob/v0.3.1/CHANGES.rst

This makes it incompatible with pyasn-modules v0.0.9. Locking pyasn1 to v0.2.3 seems to fix the issue for me.

Stack trace with pyasn1==0.3.1:

```
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/snowflake/connector/__init__.py", line 21, in <module>
    from .connection import SnowflakeConnection
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/snowflake/connector/connection.py", line 14, in <module>
    from . import network
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/snowflake/connector/network.py", line 28, in <module>
    from . import ssl_wrap_socket
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/snowflake/connector/ssl_wrap_socket.py", line 103, in <module>
    from .ocsp_pyopenssl import SnowflakeOCSP
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/snowflake/connector/ocsp_pyopenssl.py", line 37, in <module>
    from pyasn1_modules import (rfc2459, rfc2437, rfc2560)
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/pyasn1_modules/rfc2560.py", line 137, in <module>
    class ResponseData(univ.Sequence):
  File "/usr/src/app/.tox/integration-snowflake-py36/lib/python3.6/site-packages/pyasn1_modules/rfc2560.py", line 143, in ResponseData
    namedtype.NamedType('responses', univ.SequenceOf(SingleResponse())),
TypeError: __init__() takes 1 positional argument but 2 were given
```

**Comment by cmcarthur** [comment]
See #28 for fix

**Comment by smtakeda** [comment]
Yes, I've noticed this. Filed a ticket for pyasn1
https://github.com/etingof/pyasn1/issues/51
and working on the fix.

**Comment by smtakeda** [comment]
merged #28 to master. Will release 1.4.1 to pypi soon.

**Comment by smtakeda** [comment]
FYI, upgrading to `pyasn1` and `pyasn1-modules` to 0.3.1 and 0.0.10 will cause other issue:
```  File "/private/tmp/abc/lib/python3.5/site-packages/snowflake/connector/ocsp_pyopenssl.py", line 1155, in validate_by_direct_connection_simple
    ocsp_uri, ocsp_issuer, ocsp_subject)
  File "/private/tmp/abc/lib/python3.5/site-packages/snowflake/connector/ocsp_pyopenssl.py", line 1208, in validate_by_direct_connection
    raise err
  File "/private/tmp/abc/lib/python3.5/site-packages/snowflake/connector/ocsp_pyopenssl.py", line 1195, in validate_by_direct_connection
    ocsp_response, ocsp_issuer)
  File "/private/tmp/abc/lib/python3.5/site-packages/snowflake/connector/ocsp_pyopenssl.py", line 335, in process_ocsp_response
    ocsp_cert = load_certificate(FILETYPE_ASN1, cert_der)
  File "/private/tmp/abc/lib/python3.5/site-packages/OpenSSL/crypto.py", line 1661, in load_certificate
    _raise_current_error()
  File "/private/tmp/abc/lib/python3.5/site-packages/OpenSSL/_util.py", line 48, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.crypto.Error: [('asn1 encoding routines', 'ASN1_ITEM_EX_D2I', 'field missing')]
```
According to https://github.com/etingof/pyasn1/issues/54, this is due to the API changes, so those who use Python connector older than `1.4.1` must run the following commands after the python-connector is installed:
```
pip install -U pyasn1==0.2.3 pyasn1-modules==0.0.9
```

**Comment by cmcarthur** [comment]
Awesome. Thanks @smtakeda !

**Comment by csande** [comment]
Please do not use pinned versions in setup.py. If there is a concern about the latest version of pyasn1, can you please use <= or < instead of ==?

**Comment by smtakeda** [comment]
Well, this was a sort of emergency patch, and I don't see the point of having <= in this particular case, because I don't want to support any other combination of  `pyasn1` and `pyasn1-modules`, and If you look at the history of `pyasn`, 0.2.1, 0.2.2 and 0.2.3 were released within two weeks, and I don't see why 0.2.1 wanted to be used.

In general, I understand the range provides flexibility to the consumers of the package, but at the same time it will leave risk to them, and if the incident like this happens, the consumers would have to take actions anyway while the provider scrambles to address it one way or others. If it were pinned, nobody would need to panic.

The reason why I left  the versions of `pyasn` and `pyasn1-modules1` blank was just because I expected such low level API would not cause breaking change, but it happened. Obviously no enforcement of semver can be pushed, so we cannot predict what is the maximum version can guarantee the compatibility until it breaks.

I see the point using version range for `botocore`, `boto3` that release cycle is insanely frequent (well, maybe Snowflake does, though ). Getting newer versions sooner makes sense to get small or large bug fixes. But when it breaks, it breaks. All of us are taking risks.

My point is it should be balanced between pinned vs range case by case.

**Comment by csande** [comment]
By taking this approach, you might be decreasing the risk of the snowflake connector encountering incompatible behavior in these modules, but you are increasing the risk that your consumers will encounter dependency conflicts. Worse yet, it forces consumers that do encounter dependency conflicts to fork the snowflake connector (as I and others had to do in the past). If I am using another module that requires pyasn1<=0.2.2, then I cannot use that module with snowflake connector without creating a fork (even though Snowflake connector probably or possibly works with pyasn1 0.2.2). We lose all flexibility when versions are pinned in setup.py.

If we do encounter the problems that you fear, we have the option of pinning versions in our requirements.txt files. That is the proper solution for this type of problem.

Pinning versions is for requirements.txt files, not for setup.py.
